File: q08/01.rb
#!/usr/bin/env ruby

N = 12

def move(log)
  return 1 if log.size == N + 1
  cnt = 0
  [[1, 0], [-1, 0], [0, 1], [0, -1]].each do |d|
    next_pos = [log[-1][0] + d[0], log[-1][1] + d[1]]
    cnt += move(log + [next_pos]) unless log.include?(next_pos)
  end
  cnt
end

start = [0, 0]
puts move([start])

File: q08/01.swift
#!/usr/bin/swift

let n = 12

func move(log: [[Int]]) -> Int {
    guard log.count != (n + 1) else {
        return 1
    }
    var cnt = 0
    let step = [[1, 0], [-1, 0], [0, 1], [0, -1]]
    for s in step {
        var tmp = log
        let last = tmp.last!
        let next = [last[0] + s[0], last[1] + s[1]]
        if !tmp.contains(where: { $0 == next }) {
            tmp.append(next)
            cnt += move(log: tmp)
        }
    }
    return cnt
}

let start = [0, 0]
let result = move(log: [start])
print(result)
